File: java/test/Ice/operations/BatchOneways.java
// **********************************************************************
//
// Copyright (c) 2003-2013 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package test.Ice.operations;

import java.io.PrintWriter;

import test.Ice.operations.Test.MyClassPrx;
import test.Ice.operations.Test.MyClassPrxHelper;

class BatchOneways
{
    private static void
    test(boolean b)
    {
        if(!b)
        {
            throw new RuntimeException();
        }
    }

    static void
    batchOneways(MyClassPrx p, PrintWriter out)
    {
        final byte[] bs1 = new byte[10  * 1024];
        final byte[] bs2 = new byte[99  * 1024];
        final byte[] bs3 = new byte[100 * 1024];

        try
        {
            p.opByteSOneway(bs1);
        }
        catch(Ice.MemoryLimitException ex)
        {
            test(false);
        }

        try
        {
            p.opByteSOneway(bs2);
        }
        catch(Ice.MemoryLimitException ex)
        {
            test(false);
        }

        try
        {
            p.opByteSOneway(bs3);
            test(false);
        }
        catch(Ice.MemoryLimitException ex)
        {
        }

        MyClassPrx batch = MyClassPrxHelper.uncheckedCast(p.ice_batchOneway());

        for(int i = 0 ; i < 30 ; ++i)
        {
            try
            {
                batch.opByteSOneway(bs1);
            }
            catch(Ice.MemoryLimitException ex)
            {
                test(false);
            }
        }

        batch.ice_getConnection().flushBatchRequests();

        MyClassPrx batch2 = MyClassPrxHelper.uncheckedCast(p.ice_batchOneway());

        batch.ice_ping();
        batch2.ice_ping();
        batch.ice_flushBatchRequests();
        batch.ice_getConnection().close(false);
        batch.ice_ping();
        batch2.ice_ping();

        batch.ice_getConnection();
        batch2.ice_getConnection();

        batch.ice_ping();
        batch.ice_getConnection().close(false);
        try
        {
            batch.ice_ping();
            test(false);
        }
        catch(Ice.CloseConnectionException ex)
        {
        }
        try
        {
            batch2.ice_ping();
            test(false);
        }
        catch(Ice.CloseConnectionException ex)
        {
        }
        batch.ice_ping();
        batch2.ice_ping();

        Ice.Identity identity = new Ice.Identity();
        identity.name = "invalid";
        Ice.ObjectPrx batch3 = batch.ice_identity(identity);
        batch3.ice_ping();
        batch3.ice_flushBatchRequests();
        
        // Make sure that a bogus batch request doesn't cause troubles to other ones.
        batch3.ice_ping();
        batch.ice_ping();
        batch.ice_flushBatchRequests();
        batch.ice_ping();
    }
}

File: js/test/Ice/operations/BatchOneways.js
// **********************************************************************
//
// Copyright (c) 2003-2018 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

(function(module, require, exports)
{
    const Ice = require("ice").Ice;
    const test = require("TestHelper").TestHelper.test;

    async function run(communicator, prx, Test, bidir)
    {
        const bs1 = new Uint8Array(10 * 1024);
        for(let i = 0; i < bs1.length; ++i)
        {
            bs1[i] = 0;
        }

        const batch = Test.MyClassPrx.uncheckedCast(prx.ice_batchOneway());
        await batch.ice_flushBatchRequests();

        test(batch.ice_flushBatchRequests().isCompleted()); // Empty flush
        test(batch.ice_flushBatchRequests().isSent()); // Empty flush
        test(batch.ice_flushBatchRequests().sentSynchronously()); // Empty flush

        for(let i = 0; i < 30; ++i)
        {
            await batch.opByteSOneway(bs1);
        }

        let count = 0;
        while(count < 27) // 3 * 9 requests auto-flushed.
        {
            count += await prx.opByteSOnewayCallCount();
            await Ice.Promise.delay(10);
        }

        if(batch.ice_getConnection() !== null)
        {
            const batch1 = Test.MyClassPrx.uncheckedCast(prx.ice_batchOneway());
            const batch2 = Test.MyClassPrx.uncheckedCast(prx.ice_batchOneway());

            batch1.ice_ping();
            batch2.ice_ping();
            await batch1.ice_flushBatchRequests();
            await batch1.ice_getConnection().then(c => c.close(Ice.ConnectionClose.GracefullyWithWait));
            batch1.ice_ping();
            batch2.ice_ping();

            await batch1.ice_getConnection();
            await batch2.ice_getConnection();

            batch1.ice_ping();
            await batch1.ice_getConnection().then(c => c.close(Ice.ConnectionClose.GracefullyWithWait));

            batch1.ice_ping();
            batch2.ice_ping();
        }

        const identity = new Ice.Identity();
        identity.name = "invalid";
        const batch3 = batch.ice_identity(identity);
        batch3.ice_ping();
        await batch3.ice_flushBatchRequests();

        // Make sure that a bogus batch request doesn't cause troubles to other ones.
        batch3.ice_ping();
        batch.ice_ping();
        await batch.ice_flushBatchRequests();
        await batch.ice_ping();
    }
    exports.BatchOneways = {run: run};
}(typeof global !== "undefined" && typeof global.process !== "undefined" ? module : undefined,
  typeof global !== "undefined" && typeof global.process !== "undefined" ? require :
  (typeof WorkerGlobalScope !== "undefined" && self instanceof WorkerGlobalScope) ? self.Ice._require : window.Ice._require,
  typeof global !== "undefined" && typeof global.process !== "undefined" ? exports :
  (typeof WorkerGlobalScope !== "undefined" && self instanceof WorkerGlobalScope) ? self : window));
